File: src/main/java/com/arnav/tutionSAAS/controller/AdminController.java
package com.arnav.tutionSAAS.controller;

import com.arnav.tutionSAAS.dto.FeeStatusResponse;
import com.arnav.tutionSAAS.dto.InviteResponse;
import com.arnav.tutionSAAS.entity.Role;
import com.arnav.tutionSAAS.entity.User;
import com.arnav.tutionSAAS.repository.UserRepo;
import com.arnav.tutionSAAS.service.AdminService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/admin")
@PreAuthorize("hasRole('ADMIN')")
public class AdminController {

    @Autowired private AdminService adminService;
    @Autowired private UserRepo userRepo;
    @Autowired private com.arnav.tutionSAAS.service.BatchService batchService;

    // ─── Teacher Invites (link-only, no email) ───

    @PostMapping("/invites")
    public ResponseEntity<InviteResponse> generateInvite() {
        return ResponseEntity.ok(adminService.generateTeacherInvite());
    }

    @GetMapping("/invites")
    public ResponseEntity<List<InviteResponse>> getAllInvites() {
        return ResponseEntity.ok(adminService.getAllInvites());
    }

    // ─── User Listing (for dropdowns) ───

    @GetMapping("/users")
    public ResponseEntity<List<Map<String, Object>>> getUsersByRole(@RequestParam String role) {
        Role r = Role.valueOf(role.toUpperCase());
        List<User> users = userRepo.findByRole(r);
        List<Map<String, Object>> result = users.stream().map(u -> Map.<String, Object>of(
                "id", u.getId(),
                "fullName", u.getFullName() != null ? u.getFullName() : "",
                "email", u.getEmail() != null ? u.getEmail() : "",
                "grade", u.getGrade() != null ? u.getGrade() : ""
        )).collect(Collectors.toList());
        return ResponseEntity.ok(result);
    }

    @DeleteMapping("/users/{id}")
    public ResponseEntity<Void> forceDeleteUser(@PathVariable Long id) {
        User user = userRepo.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));
        userService.deleteUserByClerkId(user.getClerkId());
        return ResponseEntity.noContent().build();
    }

    // ─── Fee Management ───

    @GetMapping("/fees")
    public ResponseEntity<List<FeeStatusResponse>> getAllFeeStatus() {
        return ResponseEntity.ok(adminService.getAllFeeStatus());
    }

    @PutMapping("/students/{id}/block")
    public ResponseEntity<Void> blockStudent(@PathVariable Long id) {
        adminService.blockStudent(id);
        return ResponseEntity.noContent().build();
    }

    @PutMapping("/students/{id}/unblock")
    public ResponseEntity<Void> unblockStudent(@PathVariable Long id) {
        adminService.unblockStudent(id);
        return ResponseEntity.noContent().build();
    }

    @PutMapping("/batches/{id}/fee")
    public ResponseEntity<Void> setMonthlyFee(
            @PathVariable Long id,
            @RequestBody Map<String, Double> body) {
        double amount = body.getOrDefault("monthlyFee", 0.0);
        adminService.setMonthlyFee(id, amount);
        return ResponseEntity.noContent().build();
    }

    // ─── Batch Deletion ───

    @DeleteMapping("/batches/{id}")
    public ResponseEntity<Void> deleteBatch(@PathVariable Long id) {
        batchService.deleteBatch(id);
        return ResponseEntity.noContent().build();
    }

    // ─── Dashboard ───

    @GetMapping("/dashboard")
    public ResponseEntity<AdminService.DashboardStats> getDashboard() {
        return ResponseEntity.ok(adminService.getDashboardStats());
    }

    // ─── Join Requests ───

    @GetMapping("/join-requests")
    public ResponseEntity<List<com.arnav.tutionSAAS.dto.JoinRequestResponse>> getJoinRequests() {
        return ResponseEntity.ok(batchService.getPendingJoinRequests());
    }

    @PostMapping("/join-requests/{id}/approve")
    public ResponseEntity<Void> approveJoinRequest(@PathVariable Long id) {
        batchService.approveJoinRequest(id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/join-requests/{id}/reject")
    public ResponseEntity<Void> rejectJoinRequest(@PathVariable Long id) {
        batchService.rejectJoinRequest(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend-nest/src/admin/admin.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UseGuards,
} from '@nestjs/common';
import { Role } from '@prisma/client';
import { ClerkAuthGuard } from '../common/guards/clerk-auth.guard';
import { RolesGuard } from '../common/guards/roles.guard';
import { Roles } from '../common/decorators/roles.decorator';
import { AdminService } from './admin.service';
import { BatchesService } from '../batches/batches.service';
import { UsersService } from '../users/users.service';
import { PrismaService } from '../prisma/prisma.service';

@Controller('api/admin')
@UseGuards(ClerkAuthGuard, RolesGuard)
@Roles(Role.ADMIN)
export class AdminController {
  constructor(
    private adminService: AdminService,
    private batchesService: BatchesService,
    private usersService: UsersService,
    private prisma: PrismaService,
  ) {}

  @Post('invites')
  generateInvite() {
    return this.adminService.generateTeacherInvite();
  }

  @Get('invites')
  getAllInvites() {
    return this.adminService.getAllInvites();
  }

  @Get('users')
  async getUsersByRole(@Query('role') role: string) {
    const r = role.toUpperCase() as Role;
    const users = await this.prisma.user.findMany({ where: { role: r } });
    return users.map((u) => ({
      id: u.id,
      fullName: u.fullName || '',
      email: u.email || '',
      grade: u.grade || '',
    }));
  }

  @Delete('users/:id')
  async forceDeleteUser(@Param('id', ParseIntPipe) id: number) {
    const user = await this.prisma.user.findUnique({ where: { id } });
    if (!user) throw new Error('User not found');
    await this.usersService.deleteUserByClerkId(user.clerkId);
    return null;
  }

  @Get('fees')
  getAllFeeStatus() {
    return this.adminService.getAllFeeStatus();
  }

  @Put('students/:id/block')
  async blockStudent(@Param('id', ParseIntPipe) id: number) {
    await this.adminService.blockStudent(id);
    return null;
  }

  @Put('students/:id/unblock')
  async unblockStudent(@Param('id', ParseIntPipe) id: number) {
    await this.adminService.unblockStudent(id);
    return null;
  }

  @Put('batches/:id/fee')
  async setMonthlyFee(
    @Param('id', ParseIntPipe) id: number,
    @Body() body: { monthlyFee: number },
  ) {
    await this.adminService.setMonthlyFee(id, body.monthlyFee || 0);
    return null;
  }

  @Delete('batches/:id')
  async deleteBatch(@Param('id', ParseIntPipe) id: number) {
    await this.batchesService.deleteBatch(id);
    return null;
  }

  @Get('dashboard')
  getDashboard() {
    return this.adminService.getDashboardStats();
  }

  @Get('join-requests')
  getJoinRequests() {
    return this.batchesService.getPendingJoinRequests();
  }

  @Post('join-requests/:id/approve')
  async approveJoinRequest(@Param('id', ParseIntPipe) id: number) {
    await this.batchesService.approveJoinRequest(id);
    return null;
  }

  @Post('join-requests/:id/reject')
  async rejectJoinRequest(@Param('id', ParseIntPipe) id: number) {
    await this.batchesService.rejectJoinRequest(id);
    return null;
  }
}
